### PR TITLE
libbacktrace: Update to upstream commit 1db85642e (reverted)

### PR DESCRIPTION
**Pull request by akien-mga**
I noticed there's been a bunch of fixes upstream since 2022, so it's worth updating. Might help with some of the issues we have with dwarf-4 vs dwarf-5 on different compilers/platforms, needs testing.

https://github.com/ianlancetaylor/libbacktrace/commit/1db85642e3fca189cf4e076f840a45d6934b2456

### CODE REVIEW

File: thirdparty/libbacktrace/config.h [hunk 1]
```diff
@@ -48,7 +52,7 @@
 /* #undef HAVE_KERN_PROC_ARGS */
 
 /* Define if -llzma is available. */
-#define HAVE_LIBLZMA 1
+/* #undef HAVE_LIBLZMA */
```
Comment (akien-mga) [inline thirdparty/libbacktrace/config.h:55]: This was likely just what happened to be on bruvzg's system last time he synced.

I generated this config with:
```
./configure --host x86_64-w64-mingw32 --target x86_64-w64-mingw32 --build x86_64-w64-mingw32
```

so it should be a bit more Windows compatible by default.

Either way `HAVE_LIBLZMA` isn't used by any of the files we include.

File: thirdparty/libbacktrace/config.h [hunk 2]
```diff
@@ -95,12 +102,21 @@
...
 #define HAVE_UNISTD_H 1
 
+/* Define to 1 if you have the <windows.h> header file. */
+#define HAVE_WINDOWS_H 1
+
 /* Define if -lz is available. */
 #define HAVE_ZLIB 1
 
+/* Define if -lzstd is available. */
+/* #undef HAVE_ZSTD */
```
Comment (akien-mga) [inline thirdparty/libbacktrace/config.h:118]: Technically we have it so we could make it work, but none of the files we include make use of it.

File: thirdparty/libbacktrace/config.h [hunk 3]
```diff
@@ -95,12 +102,21 @@
...
 #define HAVE_SYS_TYPES_H 1
 
+/* Define to 1 if you have the <tlhelp32.h> header file. */
+#define HAVE_TLHELP32_H 1
+
 /* Define to 1 if you have the <unistd.h> header file. */
 #define HAVE_UNISTD_H 1
 
+/* Define to 1 if you have the <windows.h> header file. */
+#define HAVE_WINDOWS_H 1
```
Comment (akien-mga) [inline thirdparty/libbacktrace/config.h:112]: These are new, and Windows specific. Might help with some stuff?

**Comment by akien-mga** [comment]
Added some debug prints, here's what I get with dwarf-5 + gcc.
```diff
diff --git a/thirdparty/libbacktrace/read.c b/thirdparty/libbacktrace/read.c
index 7af66602fd..8b4345eb2e 100644
--- a/thirdparty/libbacktrace/read.c
+++ b/thirdparty/libbacktrace/read.c
@@ -72,9 +72,12 @@ backtrace_get_view (struct backtrace_state *state, int descriptor,
   view->len = size;
 
   got = 0;
+  printf("libbacktrace: size: %lld\n", size);
   while (got < size)
     {
+      printf("libbacktrace: size - got: %lld\n", size - got);
       r = read (descriptor, view->base, size - got);
+      printf("libbacktrace: read: %lld\n", r);
       if (r < 0)
        {
          error_callback (data, "read", errno);
```

```
libbacktrace: size: 64
libbacktrace: size - got: 64
libbacktrace: read: 64
libbacktrace: size: 24
libbacktrace: size - got: 24
libbacktrace: read: 24
libbacktrace: size: 1160
libbacktrace: size - got: 1160
libbacktrace: read: 1160
libbacktrace: size: 22284274
libbacktrace: size - got: 22284274
libbacktrace: read: 22284274
libbacktrace: size: 46764971
libbacktrace: size - got: 46764971
libbacktrace: read: 46764971
libbacktrace: size: 2234382608
libbacktrace: size - got: 2234382608
libbacktrace: read: -2060584688
[1] error(2): read
```

**Comment by akien-mga** [comment]
I got it working with both `-gdwarf-4` and `-gdwarf-5` after changing the optimization level to `-O1` to have smaller debuginfo.

My build with `-O2` has debuginfo that exceeds 2 GiB, and so it gets split forcefully into a separate 2GB `.debugsymbols` file. And so libbacktrace seems to struggle either with the file being separate, or with it being over 2GB still.

With `-O1`, the file is 1.8 GiB and since I didn't specify `separate_debug_symbols`, I have a single binary and a working stacktrace. The start of my debug prints now looks like this:
```
libbacktrace: size: 64
libbacktrace: size - got: 64
libbacktrace: read: 64
libbacktrace: size: 24
libbacktrace: size - got: 24
libbacktrace: read: 24
libbacktrace: size: 1080
libbacktrace: size - got: 1080
libbacktrace: read: 1080
libbacktrace: size: 23507590
libbacktrace: size - got: 23507590
libbacktrace: read: 23507590
libbacktrace: size: 51093710
libbacktrace: size - got: 51093710
libbacktrace: read: 51093710
libbacktrace: size: 1631000392
libbacktrace: size - got: 1631000392
libbacktrace: read: 1631000392
libbacktrace: size: 64
libbacktrace: size - got: 64
libbacktrace: read: 64
libbacktrace: size: 24
libbacktrace: size - got: 24
libbacktrace: read: 24
```
and then a lot more until:
```
libbacktrace: size: 594037
libbacktrace: size - got: 594037
libbacktrace: read: 594037
[1] __C_specific_handler (dlls/ntdll/unwind.c:2260)
[2] call_seh_handlers (dlls/ntdll/signal_x86_64.c:211)
[3] dispatch_exception (dlls/ntdll/exception.c:252)
[4] core_bind::OS::crash(String const&) (core/core_bind.cpp:274)
[5] void call_with_validated_variant_args_helper<__UnexistingClass, String const&, 0ull>(__UnexistingClass*, void (__UnexistingClass::*)(String const&), Variant const**, IndexSequence<0ull>) (./core/variant/binder_common.h:365)
[6] void call_with_validated_object_instance_args<__UnexistingClass, String const&>(__UnexistingClass*, void (__UnexistingClass::*)(String const&), Variant const**) (./core/variant/binder_common.h:652)
[7] MethodBindT<String const&>::validated_call(Object*, Variant const**, Variant*) const (./core/object/method_bind.h:357)
[8] GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Callable::CallError&, GDScriptFunction::CallState*) (modules/gdscript/gdscript_vm.cpp:2300)
[9] GDScriptInstance::callp(StringName const&, Variant const**, int, Callable::CallError&) (modules/gdscript/gdscript.cpp:2073)
[10] Node::_gdvirtual__ready_call() (scene/main/node.h:385)
[11] Node::_notification(int) (scene/main/node.cpp:227)
[12] Node::_notificationv(int, bool) (./scene/main/node.h:50)
[13] CanvasItem::_notificationv(int, bool) (./scene/main/canvas_item.h:45)
[14] Node2D::_notificationv(int, bool) (scene/2d/node_2d.h:37)
[15] Object::notification(int, bool) (core/object/object.cpp:875)
[16] Node::_propagate_ready() (scene/main/node.cpp:282)
[17] Node::_propagate_ready() (scene/main/node.cpp:273)
[18] Node::_set_tree(SceneTree*) (scene/main/node.cpp:3259)
[19] SceneTree::initialize() (scene/main/scene_tree.cpp:480)
[20] OS_Windows::run() (platform/windows/os_windows.cpp:1995)
[21] widechar_main(int, wchar_t**) (platform/windows/godot_windows.cpp:180)
[22] _main() (platform/windows/godot_windows.cpp:206)
[23] main (platform/windows/godot_windows.cpp:225)
[24] __tmainCRTStartup (../crt/crtexe.c:267)
[25] WinMainCRTStartup (../crt/crtexe.c:157)
[26] BaseThreadInitThunk (dlls/kernel32/thread.c:61)
-- END OF BACKTRACE --
```

Could it be that our issue with dwarf-4/dwarf-5 is just that we happen to be close to the 2.0 GiB int32 limit and that's causing issues when we tip over?

**Review by hpvb — APPROVED** [review]
I have ran this for a little bit, and on Linux it seems to work fine. You already did some mingw checks so I think this is good.

**Comment by akien-mga** [comment]
Note that we're not using libbacktrace for Linux currently, only for Windows/mingw.

We could, though, if there's benefit to it compared to our current code.

**Comment by bruvzg** [comment]
> Note that we're not using libbacktrace for Linux currently, only for Windows/mingw.

If it's working reliably (was not when I last tested, but it was a long time ago), the only benefit probably is getting rid of dependency on external `addr2line` (or `atos` on macOS, but it's less relevant, I think it's part of base system install).

**Comment by akien-mga** [comment]
`addr2line` is part of binutils which should be relatively ubiquitous on Linux setups that are suitable for gaming or gamedev. But it's true that the dependency _can_ be missing, e.g. it's not preinstalled in Fedora's minimal docker image (but that image is kept as minimal as possible by design).

I don't think it's been an issue up until now so I'd say let's not fix what isn't broken, and keep the current implementation.

**Comment by hpvb** [comment]
> Note that we're not using libbacktrace for Linux currently, only for Windows/mingw.
> 
> We could, though, if there's benefit to it compared to our current code.

Well then, probably should still upgrade. 😅😅😅

**Comment by Repiteo** [comment]
Thanks!

**Comment by akien-mga** [comment]
Reverted by #102900.

**Pull request merged**